tests/unit/test_charm.py

```diff
@@ -26,7 +42,7 @@ def test_pebble_container_cannot_connect(harness: Harness) -> None:
     assert harness.charm.unit.status == WaitingStatus("Waiting to connect to Oathkeeper container")
 
 
-def test_update_container_config(harness: Harness) -> None:
+def test_update_container_config_without_kratos_relation(harness: Harness) -> None:
```

Comment:
I am thinking if we could simplify these two test cases `test_update_container_config_without_kratos_relation` and `test_update_container_config_with_kratos_relation`. It looks like pasting all of the configuration content for the `expected_config` is not necessary and hard to maintain. I guess it would be fine to just use `template.render` to prepare the `expected_config` variable is good enough.

Reply:
I understand your point, but I'm not sure that will cover what we want to test.
If we use `harness.charm._render_conf_file()` for `expected_config` and then compare it against what was pushed to the container, it'll be like checking if 2 exact things are equal (since we push the output of render method).
Instead, we want to see if the pushed config file looks like expected, which at the same time tests that the render method works correctly.

To give a simple example, right now the test would catch if the following:
```yaml
check_session_url: {{ kratos_session_url | d("http://default-kratos-url/sessions/whoami", true) }}
```
was rendered correctly to:
```yaml
check_session_url: "some-kratos-relation-url"
```
if kratos relation was in place.

Without having the specified expected content of the config, the test wouldn't catch if the config rendered the default "http://default-kratos-url/sessions/whoami" instead, or if some matching conditional yaml block was added. I hope that explains my reasoning.